### PR TITLE
Do not run the 'dependency-submission' job in CI for forked repositories

### PR DESCRIPTION
**Pull request by danicheg**
This tries to fix the following issue in forked repositories:
```
[error] Unexpected status 404 Not Found with body:
[error] {"message":"The Dependency graph is disabled for this repository. Please enable it before submitting snapshots.","documentation_url":"https://docs.github.com/rest/dependency-graph/dependency-submission#create-a-snapshot-of-dependencies-for-a-repository","status":"404"}
```
Example: https://github.com/danicheg/cats/actions/runs/10027887278/job/27714122616#step:10:554

### CODE REVIEW

**Comment by danicheg** [comment]
It looks like it works https://github.com/danicheg/sbt-typelevel/actions/runs/10043935348/job/27757804649

**Review by armanbilge — COMMENTED** [review]
Thanks!

This is something I've thought about doing before. We have a similar problem not just with dependency submission but also with publishing: forks also attempt to publish even though their credentials are not setup correctly. So we should also consider fixing that in a similar way.

Although this solution works well for CI, the reason I've hesitated to implement it before is because it requires that the developer's local git repository is correctly configured with the `upstream` remote. Unfortunately, this does not happen automatically: if you fork a repo on GitHub and then `git clone` it locally, the `upstream` will not be configured and until you do so you will not be able to properly generate workflows.

Thoughts?

**Comment by danicheg** [comment]
> it requires that the developer's local git repository is correctly configured with the upstream remote. Unfortunately, this does not happen automatically: if you fork a repo on GitHub and then git clone it locally, the upstream will not be configured and until you do so you will not be able to properly generate workflows.

Oh, I see. Several thoughts:
1. Because we have a CI check, the issue with the wrong Github Repository Owner will be clearly detected. So, a user might manually fix the issue (it's a one-time action) or configure the git repository properly to meet our requirements.
2. We can have a distinct `tlGitHubRepoOwner` SBT setting. Users will have the ability to explicitly set it. Or it will be filled in via `GitHelper#extractGitHubUserRepo`.

File: ci/src/main/scala/org/typelevel/sbt/TypelevelCiPlugin.scala [hunk 1]
```diff
@@ -140,7 +141,13 @@ object TypelevelCiPlugin extends AutoPlugin {
     githubWorkflowJavaVersions := Seq(JavaSpec.temurin("8")),
     githubWorkflowAddedJobs ++= {
       val dependencySubmission =
-        if (tlCiDependencyGraphJob.value)
+        if (tlCiDependencyGraphJob.value) {
+          val githubRepo = GitHelper.extractGitHubUserRepo.map(x => x._1 + '/' + x._2)
+          val jobCondition =
+            githubRepo.fold(
+              "github.event_name != 'pull_request'"
+            )(repo => s"github.repository == '$repo' && github.event_name != 'pull_request'")
```
[the remark this answers is not in the document]
Comment (danicheg) [reply ci/src/main/scala/org/typelevel/sbt/TypelevelCiPlugin.scala:149]: yeah, it makes sense 👍🏻

**Comment by danicheg** [comment]
> Also, if it really bothers users, why don't they take the "one-time action" to enable Dependency Graph for their repository? Wouldn't that solve the problem?

There seems to be some confusion when we talk about 'users.' From my perspective, a typical user of sbt-typelevel is a maintainer of an affiliate project. In 99% of cases, these users are the ones who set up CI workflows in their projects. We can reasonably expect them to have decent skills in a wide range of areas, such as setting up a local git repository. On the other hand, there are users of these affiliate projects. These users are more likely to focus on new features, bug fixing, and enhancing documentation in their typical contributions, rather than adjusting the CI workflow. However, these users might encounter confusion in their forks when running CI workflows. To sum up:
This work aims to improve the DX of contributors (by reducing the confusion caused by failing CI in their forks) of affiliate projects, rather than focusing on the maintainers of these projects. Let's be honest with ourselves, nowadays we can count them on one hand.

**Comment by armanbilge** [comment]
> On the other hand, there are users of these affiliate projects. These users are more likely to focus on new features, bug fixing, and enhancing documentation in their typical contributions, rather than adjusting the CI workflow.

Ah yes, I see your point. I agree.

I do have one last idea, that would make workflow generation independent of the git repository configuration, by using a condition like this:
```
if: github.event.repository.fork == false
```

I think what we should do is serialize this as a setting e.g. `tlCiForkCondition` with that as the default value. In the unlikely situation that the primary repository is a fork (e.g. of some unmaintained upstream) then they can reconfigure it appropriately.

WDYT?

**Comment by danicheg** [comment]
If you argue that this approach feels better, it's okay with me. I'd like to have that issue resolved, no matter the path we take.

**Comment by danicheg** [comment]
@armanbilge by the way, If I understand things correctly, your suggested solution won't handle the case when the affiliate project is a fork. `tlCiForkCondition` is set to `true`, `github.event.repository.fork` would be computed as `true` for users' forks — and we'll get the run `dependency-submission` when it's unsolicited. Should we worry about this case? Perhaps you know better 🤷🏻 I'm not contributing to such projects within Typelevel.

**Comment by armanbilge** [comment]
> If I understand things correctly, your suggested solution won't handle the case when the affiliate project is a fork

By default, no. That is why I suggested we implement it as a configurable setting:

> I think what we should do is serialize this as a setting e.g. `tlCiForkCondition` with that as the default value. In the unlikely situation that the primary repository is a fork (e.g. of some unmaintained upstream) then they can reconfigure it appropriately.

Repositories that are maintained as forks are very unusual/rare.

**Comment by danicheg** [comment]
@armanbilge So, is there anything I need to address here to move on with this?

**Review by armanbilge — COMMENTED** [review]
Yes, sorry. I think your concern was legitimate, which is why I suggested we implement this as a setting https://github.com/typelevel/sbt-typelevel/pull/720#issuecomment-2254592294.

**Comment by danicheg** [comment]
Feeling dumb, but I altered the initial approach to align with your suggestions. So now we have a configurable setting with a reasonable default value 🤔

**Comment by armanbilge** [comment]
So sorry, I think I didn't explain very well. What I mean is that we should have `String` setting `tlCiForkCondition` which has the default value `github.event.repository.fork == false`, but if necessary users can customize it to `github.repository == '$repo'`.

**Comment by danicheg** [comment]
Returning to this... 

> What I mean is that we should have String setting tlCiForkCondition which has the default value github.event.repository.fork == false, but if necessary users can customize it to github.repository == '$repo'.

Wouldn't that `tlCiForkCondition` setting be somewhat of a leaky abstraction? I mean, it wouldn't just specify some behaviour of the underlying GitHub Action but would actually allow the injection of any YAML into the GitHub Action. I don't know if this is already the usual practice in the plugin, though.

**Comment by armanbilge** [comment]
> but would actually allow the injection of any YAML into the GitHub Action

Well, not _any_ YAML, just a conditional expression. And we already accept an arbitrary string anyway.

https://github.com/typelevel/sbt-typelevel/blob/e3849909cfa076127be6fddb504827dad1facb0e/github-actions/src/main/scala/org/typelevel/sbt/gha/WorkflowStep.scala#L22

Any plugin could always manipulate the `gitHubWorkflowSteps` setting and do arbitrary things to the `cond`s.

**Comment by danicheg** [comment]
Okay, as I said, any solution would work for me if it resolves the original request.

**Comment by danicheg** [comment]
@armanbilge hey Arman, what's up? Is there anything else I need to polish to move forward with this?

**Review by armanbilge — APPROVED** [review]
Sorry, just got a bit behind :) this looks good, thanks for adjusting!

**Pull request merged**